lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java

```diff
@@ -101,4 +115,207 @@ public void leaderIsInitialisedForAllMethods() {
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 15L, LEADER_2),
+                ImmutableSingleCall.of(21L, 28L, LEADER_2),
+                ImmutableSingleCall.of(36L, 45L, LEADER_3));
+        assertExpectedDuration(Instant.ofEpochMilli(21L), Instant.ofEpochMilli(45L));
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *                [ B ]
+     *             [    A    ]
+     *             <======>
+     */
+    @Test
+    public void updateOldLeaderUpperBoundWithNewRequests() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(15L, 21L, LEADER_2),
+                ImmutableSingleCall.of(10L, 28L, LEADER_1));
+        assertExpectedDuration(Instant.ofEpochMilli(10L), Instant.ofEpochMilli(21L));
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [ B ]
+     *                   [ C ]
+     *                         [ C ]
+     *       <=========>
+     */
+    @Test
+    public void detectFirstLeaderElectionFromLongTermLeader() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 15L, LEADER_2),
+                ImmutableSingleCall.of(21L, 28L, LEADER_3),
+                ImmutableSingleCall.of(36L, 45L, LEADER_3));
+        assertExpectedDuration(Instant.ofEpochMilli(3L), Instant.ofEpochMilli(15L));
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [       B       ]
+     *                [ C ]
+     *                      [ C ]
+     *       <============>
+     */
+    @Test
+    public void detectFirstLeaderElectionFromLongTermLeaderWithSlowRequest() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 45L, LEADER_2),
+                ImmutableSingleCall.of(15L, 21L, LEADER_3),
+                ImmutableSingleCall.of(28L, 36L, LEADER_3));
+        assertExpectedDuration(Instant.ofEpochMilli(3L), Instant.ofEpochMilli(21L));
+    }
+
+    /**
+     * [ A ]
+     *       [        B       ]
+     *          [ C ]
+     *                 [ C ]
+     */
+    @Test
+    public void doNotCalculateLeadershipBeforeFirstLongTermLeader() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 36L, LEADER_2),
+                ImmutableSingleCall.of(10L, 15L, LEADER_3),
+                ImmutableSingleCall.of(21L, 28L, LEADER_3));
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isEmpty();
+    }
+
+    /**
+     * [ A ]
+     *       [        A       ]
+     *          [ C ]
+     *                 [ C ]
+     *       <======>
```

Comment:
this is clever. Thanks a lot for the ASCII art, really helps convey what's going on